### DISCUSSION TITLE
How to refresh stream after some times?

### DISCUSSION THREAD

**Discussion by d222gfxx** [streamlink/streamlink#4165] [Q&A] [answered]
I am recording a website's live stream, HLSStream keep got 404 error after 1 hour, so I need to refresh stream every half hour to get an uninterruptible full stream.

[https://github.com/streamlink/streamlink/issues/2626](url)
Conclusion here is this function should be handled plugin specified, but I am not sure how to write code, is there any code snippet I can take a look?

And I am curious, is this expired thing only happened in HLSStream? Or it can happend in RTMPStream HTTPStream DashStream too?

**Comment by bastimeyer** [comment, marked as answer]
> HLSStream 404

If the HLSStream URL expires but can be re-acquired, then you can turn the `HLSStream.url` attribute into a property method which recomputes the URL when it gets accessed. The worker thread will read the `url` attribute whenever it reloads the playlist to queue new segments.

Recent example here:
https://github.com/streamlink/streamlink/blob/47b715a3693171cb6e4ee2968627770bb3d5c1bc/src/streamlink/plugins/oneplusone.py#L37-L55

> And I am curious, is this expired thing only happened in HLSStream?

Resource expiration can happen regardless of the streaming protocol. This entirely depends on the streaming service. If a streaming service feels like they don't have to follow a well established protocol and want to do extra things or things which are entirely different, then it's up to them.

Another recent example of custom HLS logic where playlists and segments require URL query string updates, which is off-spec:
https://github.com/streamlink/streamlink/blob/47b715a3693171cb6e4ee2968627770bb3d5c1bc/src/streamlink/plugins/ltv_lsm_lv.py#L19-L47
